src/core/job.c

```diff
@@ -1668,6 +1611,16 @@ int job_compare(Job *a, Job *b, UnitDependency assume_dep) {
         if (a->ignore_order || b->ignore_order)
                 return 0;
 
+        /* Figure out the dependency between units */
+        if (assume_dep == 0) {
+                if (hashmap_contains(ua->dependencies[UNIT_AFTER], ub))
+                        assume_dep = UNIT_AFTER;
+                /* We should not be called with a loop between jobs, let UNIT_BEFORE override not to fail. */
+                if (hashmap_contains(ub->dependencies[UNIT_AFTER], ua))
```

Comment:
Related to the loop between jobs below (if/else if decides whether to what dependency we fall back, it shouldn't matter).

Reply:
Sorry, I wasn't clear. Only one of the two conditions should be possible (either `hashmap_contains(ua->dependencies[UNIT_AFTER], ub)` or `hashmap_contains(ub->dependencies[UNIT_AFTER], ua)`), (and the second assignment overrides the first assignment anyway), so `else if` should be used to avoid an unnecessary second lookup in a hashmap.

Reply:
Good point about the redundant lookup. Fixed.